packages/block/src/blockchainMachine.ts

```diff
@@ -34,82 +36,125 @@ export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopServi
                 await component.applyAction(a.value);
                 await this.actionStore.removeItem(component.name, a);
             } catch (doh) {
-                this.logger.error(doh);
+                logger.error(doh);
             }
         });
     }
 
-    protected async startInternal(): Promise<void> {
-        if (!this.blockProcessor.started) this.logger.error("The BlockProcessor should be started before the BlockchainMachine.");
-        if (!this.actionStore.started) this.logger.error("The actionStore should be started before the BlockchainMachine.");
-        if (!this.blockItemStore.started) this.logger.error("The BlockItemStore should be started before the BlockchainMachine.");
-
-        this.blockProcessor.newBlock.addListener(this.processNewBlock);
-
-        // For each component, load and start any action that was stored in the actionStore
+    /**
+     * Finds any existing actions in the store and executes them.
+     */
+    public executeExistingActions() {
         for (const component of this.components) {
             const actionAndIds = this.actionStore.getItems(component.name);
             this.runActionsForComponent(component, actionAndIds);
         }
     }
 
-    protected async stopInternal(): Promise<void> {
-        this.blockProcessor.newBlock.removeListener(this.processNewBlock);
-    }
-
-    constructor(
-        private blockProcessor: BlockProcessor<TBlock>,
-        private actionStore: CachedKeyValueStore<ComponentAction>,
-        private blockItemStore: BlockItemStore<TBlock>
-    ) {
-        super("blockchain-machine");
-        this.processNewBlock = this.processNewBlock.bind(this);
-    }
-
     /**
-     * Add a new `component` to the BlockchainMachine. It must be called before the service is started
-     * @param component
+     * Sets a state for a block whose parent may not already be in the block item store.
+     * If the parent is in the store reduce is used to compute the next state, if it is not the getInitialState is
+     * called on the component reducer.
+     * This function writes to the blockItemStore and MUST be included in a batch to persist beyond memory.
+     * @param block
      */
-    public addComponent(component: Component<AnchorState, TBlock, ComponentAction>): void {
-        if (this.started) throw new ApplicationError("Components must be added before the BlockchainMachine is started.");
-        if (this.componentNames.has(component.name)) throw new ApplicationError(`A Component with the name "${component.name}" was already added.`);
+    public async setInitialState(block: TBlock): Promise<void> {
+        if (!this.actionStore.started) logger.error("The actionStore should be started before the BlockchainMachine.");
+        if (!this.blockItemStore.started) logger.error("The BlockItemStore should be started before the BlockchainMachine.");
```

Comment:
The error messages don't make too much sense now that this is separated from the `BlockchainMachineService`, as the BlockchainMachine is not really "started". Same below in `setStateAndDetectChanges`.

Reply:
fixed